Src/Plugins/Security/SEC101_054.AlibabaAccessKeyValidator.cs

```diff
@@ -0,0 +1,119 @@
+                                 ref groups,
+                                 ref failureLevel,
+                                 ref fingerprint,
+                                 ref message);
+        }
+
+        public static string IsValidDynamic(ref string fingerprint, ref string message)
+        {
+            return IsValidDynamic(Instance,
+                                   ref fingerprint,
+                                   ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("account", out string account) ||
+                !groups.TryGetNonEmptyValue("password", out string password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Platform = nameof(AssetPlatform.AlibabaCloud),
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            var fingerprint = new Fingerprint(fingerprintText);
+            string account = fingerprint.Account;
+            string password = fingerprint.Password;
+
+            try
+            {
+                // Taken from https://www.alibabacloud.com/help/doc-detail/63638.htm
+                IClientProfile clientProfile = DefaultProfile.GetProfile("cn-shanghai", account, password);
+                DefaultAcsClient client = new DefaultAcsClient(clientProfile);
+
+                PubRequest request = new PubRequest();
+
+                // We don't actually care about any of the product stuff, so just leave dummy values in
+                request.ProductKey = "<productKey>";
+                request.TopicFullName = "/<productKey>/<deviceName>/get";
+                byte[] payload = Encoding.Default.GetBytes("Hello World.");
+                string payloadStr = Convert.ToBase64String(payload);
+                request.MessageContent = payloadStr;
+                request.Qos = 0;
+                PubResponse response = client.GetAcsResponse(request);
+
+                // If all goes well, we'll receive a "product format invalid" message in the response
+                // which means authentication succeeded. Therefore the id and secret are valid.
+                return ReturnAuthorizedAccess(ref message, asset: account);
```

Comment:
>ReturnAuthorizedAccess [](start = 23, length = 22)

Generally, we have placed these at the very exit point at the end of the function. i.e., look at other rules and keep your implementation consistent w/the other patterns.

Reply:
Done.

---
In reply to: [591882758](https://github.com/microsoft/sarif-pattern-matcher/pull/285#discussion_r591882758) [](ancestors = 591882758)